validate.swift

```diff
@@ -2,67 +2,450 @@
 
 import Foundation
 
-// Find the "packages.json" file based on arguments, current directory, or the directory of the script
-let argumentURL = CommandLine.arguments.dropFirst().first.flatMap(URL.init(fileURLWithPath:))
-let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath, isDirectory: true).appendingPathComponent("packages.json")
-let scriptDirectoryURL = URL(fileURLWithPath: #file).deletingLastPathComponent().appendingPathComponent("packages.json")
+// MARK: Configuration Values and Constants
+
+// number of validations to run simultaneously
+let semaphoreCount = 12
+
+let timeoutIntervalForRequest = 3000.0
+let timeoutIntervalForResource = 6000.0
+
+// base url for github raw files
+let rawURLComponentsBase = URLComponents(string: "https://raw.githubusercontent.com")!
+
+// master package list to compare against
+let masterPackageList = rawURLComponentsBase.url!.appendingPathComponent("daveverwer/SwiftPMLibrary/master/packages.json")
 
-let possibleURLs : [URL?] = [argumentURL, currentDirectoryURL, scriptDirectoryURL]
+let helpText = """
+usage: %@ <command> [path]
 
-guard let url = possibleURLs.compactMap({ $0 }).first( where: { FileManager.default.fileExists(atPath: $0.path )}) else {
-  print("Error: Unable to find packages.json to validate.")
-  exit(1)
+COMMANDS:
+  all   validate all packages in JSON packages.json
+  diff  validate all new packages in JSON packages.json
+  mine  validate the Package of the current directoy
+
+OPTIONS:
+  path  to define the specific `packages.json` file or Swift package directory
+"""
+// MARK: Types
+
+
+enum Command : String {
+  case all = "all"
+  case diff = "diff"
+  case mine = "mine"
 }
 
-let decoder = JSONDecoder()
-let data = try! Data(contentsOf: url)
-let packageUrls  = try! decoder.decode([URL].self, from: data)
+extension Command {
+  static func fromArguments (_ arguments : [String]) -> Command? {
```

Comment:
That's a little bizarre. We are passing an array of arguments and we are returning one command. But we are initialising the command with only on of the arguments(potentials the first valid one). Is that expected?

Reply:
Yes. `all` `mine` and `diff` are exclusive. I just take the first valid subcommand. I am open to suggestions though.